lib/chef/knife/cloud/vra_service.rb

```diff
@@ -0,0 +1,139 @@
+  class Knife
+    class Cloud
+      class VraService < Service
+        include VraServiceHelpers
+        def initialize(options={})
+          super(options)
+
+          @username   = options[:username]
+          @password   = options[:password]
+          @base_url   = options[:base_url]
+          @tenant     = options[:tenant]
+          @verify_ssl = options[:verify_ssl]
+        end
+
+        def connection
+          @client ||= Vra::Client.new(
+            base_url:   @base_url,
+            username:   @username,
+            password:   @password,
+            tenant:     @tenant,
+            verify_ssl: @verify_ssl
+          )
+        end
+
+        def create_server(options={})
+          submitted_request = catalog_request(options).submit
+          ui.msg("Catalog request #{submitted_request.id} submitted.")
+          wait_for_request(submitted_request, options[:wait_time], options[:refresh_rate])
+          ui.msg("Catalog request complete.\n")
+          request_summary(submitted_request)
+
+          raise CloudExceptions::ServerCreateError, submitted_request.completion_details if submitted_request.failed?
+
+          servers = submitted_request.resources.select(&:vm?)
+          raise CloudExceptions::ServerCreateError, 'The vRA request created more than one server, but we were only expecting 1' if servers.length > 1
+          raise CloudExceptions::ServerCreateError, 'The vRA request did not create any servers' if servers.length == 0
+
+          servers.first
+        end
+
+        def delete_server(instance_id)
+          server = get_server(instance_id)
+          server_summary(server)
+          ui.msg('')
+
+          if server.status == 'DELETED'
+            ui.warn('Server is already deleted.')
+            ui.msg('')
+            return
+          end
+
+          ui.confirm('Do you really want to delete this server')
+
+          destroy_request = server.destroy
+          ui.msg("Destroy request #{destroy_request.id} submitted.")
+          wait_for_request(destroy_request)
+          ui.msg('Destroy request complete.')
+          request_summary(destroy_request)
+        end
+
+        def list_servers
+          connection.resources.all_resources.select(&:vm?)
+        end
+
+        def list_catalog_items(entitled)
+          if entitled
+            connection.catalog.entitled_items
+          else
+            connection.catalog.all_items
+          end
+        end
+
+        def get_server(instance_id)
+          connection.resources.by_id(instance_id)
+        end
+
+        def server_summary(server, _columns_with_info=nil)
+          msg_pair('Server ID', server.id)
+          msg_pair('Server Name', server.name)
+          msg_pair('IP Addresses', server.ip_addresses.nil? ? 'none' : server.ip_addresses.join(', '))
+          msg_pair('Status', server.status)
+          msg_pair('Catalog Name', server.catalog_name)
+        end
+
+        def request_summary(request)
+          ui.msg('')
```

Comment:
They are intended for cleanliness in STDOUT.  Otherwise, it looks like a big jumbled mess.  The blank lines separate out relevant sections of output.